hydra-cluster/src/Hydra/Cluster/Faucet.hs

```diff
@@ -60,47 +81,37 @@ seedFromFaucet ::
   Marked ->
   Tracer IO FaucetLog ->
   IO UTxO
-seedFromFaucet RunningNode{networkId, nodeSocket} receivingVerificationKey lovelace marked tracer = do
+seedFromFaucet cardanoNode@RunningNode{networkId, nodeSocket} receivingVerificationKey lovelace marked tracer = do
   (faucetVk, faucetSk) <- keysFor Faucet
-  retryOnExceptions $ submitSeedTx faucetVk faucetSk
+  retryOnExceptions tracer $
+    buildAndSubmitTx cardanoNode lovelace marked receivingAddress faucetVk faucetSk True
   waitForPayment networkId nodeSocket lovelace receivingAddress
  where
-  isResourceExhausted ex = case ioe_type ex of
-    ResourceExhausted -> True
-    _ -> False
-
-  retryOnExceptions action =
-    action
-      `catches` [ Handler $ \(_ :: SubmitTransactionException) -> do
-                    threadDelay 1
-                    retryOnExceptions action
-                , Handler $ \(ex :: IOException) -> do
-                    unless (isResourceExhausted ex) $
-                      throwIO ex
-                    traceWith tracer $
-                      TraceResourceExhaustedHandled $
-                        "Expected exception raised from seedFromFaucet: " <> show ex
-                    threadDelay 1
-                    retryOnExceptions action
-                ]
-
-  submitSeedTx faucetVk faucetSk = do
-    faucetUTxO <- findUTxO faucetVk
-    let changeAddress = ShelleyAddressInEra (buildAddress faucetVk networkId)
-    buildTransaction networkId nodeSocket changeAddress faucetUTxO [] [theOutput] >>= \case
-      Left e -> throwIO $ FaucetFailedToBuildTx{reason = e}
-      Right body -> do
-        submitTransaction networkId nodeSocket (sign faucetSk body)
-
-  findUTxO faucetVk = do
-    faucetUTxO <- queryUTxO networkId nodeSocket QueryTip [buildAddress faucetVk networkId]
-    let foundUTxO = UTxO.filter (\o -> txOutLovelace o >= lovelace) faucetUTxO
-    when (null foundUTxO) $
-      throwIO $ FaucetHasNotEnoughFunds{faucetUTxO}
-    pure foundUTxO
-
   receivingAddress = buildAddress receivingVerificationKey networkId
 
+buildAndSubmitTx ::
+  RunningNode ->
+  -- | Amount of lovelace to send to the reciver
+  Lovelace ->
+  -- | Marked as fuel or normal output?
+  Marked ->
+  -- | Receiving address
+  Address ShelleyAddr ->
+  -- | Sender verification key
+  VerificationKey PaymentKey ->
+  -- | Sender signing key
+  SigningKey PaymentKey ->
+  -- | Should we filter utxo to find the one containing more lovelace than we want to send?
+  Bool ->
```

Comment:
Should: not compose functions this way.

This `Bool` is a smell to me. It indicates that you are using only a part of the code below in some cases. Why not structure it differently to assemble what you need and don't need better?

I'll point out more comments at the usage sites.